pennylane/templates/subroutines/trotter.py

```diff
@@ -0,0 +1,284 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Contains templates for Suzuki-Trotter approximation based subroutines.
+"""
+import pennylane as qml
+from pennylane.operation import Operation
+from pennylane.ops import Sum
+
+
+def _scalar(order):
+    """Compute the scalar used in the recursive expression.
+
+    Args:
+        order (int): order of Trotter product (assume order is an even integer > 2).
+
+    Returns:
+        float: scalar to be used in the recursive expression.
+    """
+    root = 1 / (order - 1)
+    return (4 - 4**root) ** -1
+
+
+@qml.QueuingManager.stop_recording()
+def _recursive_expression(x, order, ops):
+    """Generate a list of operations using the
+    recursive expression which defines the Trotter product.
+
+    Args:
+        x (complex): the evolution 'time'
+        order (int): the order of the Trotter Expansion
+        ops (Iterable(~.Operators)): a list of terms in the Hamiltonian
+
+    Returns:
+        List: the approximation as product of exponentials of the Hamiltonian terms
+    """
+    if order == 1:
+        return [qml.exp(op, x * 1j) for op in ops]
+
+    if order == 2:
+        return [qml.exp(op, x * 0.5j) for op in ops + ops[::-1]]
+
+    scalar_1 = _scalar(order)
+    scalar_2 = 1 - 4 * scalar_1
+
+    ops_lst_1 = _recursive_expression(scalar_1 * x, order - 2, ops)
+    ops_lst_2 = _recursive_expression(scalar_2 * x, order - 2, ops)
+
+    return (2 * ops_lst_1) + ops_lst_2 + (2 * ops_lst_1)
+
+
+class TrotterProduct(Operation):
+    r"""An operation representing the Suzuki-Trotter product approximation for the complex matrix
+    exponential of a given Hamiltonian.
+
+    The Suzuki-Trotter product formula provides a method to approximate the matrix exponential of
+    Hamiltonian expressed as a linear combination of terms which in general do not commute. Consider
+    the Hamiltonian :math:`H = \Sigma^{N}_{j=0} O_{j}`, the product formula is constructed using
+    symmetrized products of the terms in the Hamiltonian. The symmetrized products of order
+    :math:`m \in [1, 2, 4, ..., 2k] | k \in \mathbb{N}` are given by:
```

Comment:
Minor, but the `| k` part looks a bit compressed in the docs, maybe we say "with $k \in \mathbb{N}$"